llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -7549,16 +7549,20 @@ createWidenInductionRecipes(PHINode *Phi, Instruction *PhiOrTrunc,
   assert(SE.isLoopInvariant(IndDesc.getStep(), &OrigLoop) &&
          "step must be loop invariant");
 
+  // It is always safe to copy over the NoWrap and FastMath flags. In
+  // particular, when folding tail by masking, the masked-off lanes are never
+  // executed, so it is safe.
```

Comment:
To clarify, the lanes are still excecuted, but they are not used, as they are masked off.

I am not sure if that holds in all cases, one case to double check is: we have a wide canonical induction that has is used to compute the header mask. In that case, none of its lanes get masked. I think the current runtime checks checks ensure that the it won't unsigned-wrap, but it may sign-wrap, poisoning if it has NSW flags.

This is not an issue for floating point inductions.

Reply:
Yeah, I'm also confused by this: where would VPWidenCanonicalIVRecipe would appear? How is it related to VPWidenIntOrFPInductionRecipe, which is a VPWidenInductionRecipe?

Reply:
We start out with VPWidenCanonicalIVRecipe if it is needed, and `removeRedundantCanonicalIVs` tries to replace them with a suitable wide induction, if there is one.

For this case, it should be fine to drop NSW, not sure if there are other cases that could cause similar problems, might be good to check if there's anything we can verify?